### PR TITLE
feat: Enable range queries for secondary indexes

### LINKED ISSUE #3651: Sec. index: make sec. indexes take advantage of range querier to KV datastore
Before we switch to new core KV storage there was no option to run range queries, so the only reasonable queries sec. indexes could use was prefix queries. This means that for filters like `_gt`, `_lt` and others we need to scan all values for the index and compare.

Now that we have range queries, we can use them to optimize comparison queries.

### PR DESCRIPTION
**Pull request by islamaliev**
## Relevant issue(s)

Resolves #3651

## Description

Make secondary indexes use range capabilities of corekv storage.

### CODE REVIEW

**Review by AndrewSisley — APPROVED** [review]
LGTM, just a few places that I think should be clarified before merge, thanks Islam :)

File: internal/db/fetcher/indexer_iterators.go [hunk 1]
```diff
@@ -236,12 +256,31 @@ func (iter *inIndexIterator) nextIterator() (bool, error) {
...
 	}
 
-	switch fieldIter := iter.indexIterator.(type) {
-	case *indexPrefixIterator:
-		fieldIter.indexKey.Fields[0].Value = iter.inValues[iter.nextValIndex]
-	case *eqSingleIndexIterator:
-		fieldIter.indexKey.Fields[0].Value = iter.inValues[iter.nextValIndex]
+	if iter.isUnique {
+		keyFieldValues := make([]client.NormalValue, len(iter.fieldConditions))
+		keyFieldValues[0] = iter.inValues[iter.nextValIndex]
```
Comment (shahzadlone) [inline internal/db/fetcher/indexer_iterators.go:261]: question: Should there be a check to ensure if it's of length 1 at least (i.e. could it panic with `0` ever?)

**Review by shahzadlone — APPROVED** [review]
LGTM :)

File: internal/db/fetcher/indexer_iterators.go [hunk 2]
```diff
@@ -162,19 +178,19 @@ func (f *indexFetcher) newPrefixIterator(
 	indexKey keys.IndexDataStoreKey,
 	matchers []valueMatcher,
 	execInfo *ExecInfo,
-) *indexPrefixIterator {
-	return &indexPrefixIterator{
+) *indexMatchIterator {
```
Comment (fredcarle) [inline internal/db/fetcher/indexer_iterators.go:181]: suggestion: The name of the method should reflect the outcome. In this case it should probably be `newMatchIterator`.

File: internal/db/fetcher/indexer_iterators.go [hunk 3]
```diff
@@ -337,7 +376,7 @@ func (iter *memorizingIndexIterator) Close() error {
 func (f *indexFetcher) newPrefixIteratorFromConditions(
 	fieldConditions []fieldFilterCond,
 	matchers []valueMatcher,
-) (*indexPrefixIterator, error) {
+) (*indexMatchIterator, error) {
```
Comment (fredcarle) [inline internal/db/fetcher/indexer_iterators.go:379]: suggestion: Here again the name of the method should reflect the outcome `newMatchIteratorFromConditions`

File: internal/db/fetcher/indexer_iterators.go [hunk 4]
```diff
@@ -440,6 +493,147 @@ func (f *indexFetcher) newIndexDataStoreKeyWithValues(values []client.NormalValu
 	return keys.NewIndexDataStoreKey(shortID, f.indexDesc.ID, fields), nil
 }
 
+// incrementKeyBytes increments a key by appending bytes to create the next possible key boundary.
+// This works because keys are compared lexicographically.
+func incrementKeyBytes(key []byte) []byte {
+	// For non-unique indexes, we need to account for the document ID that comes after the value
+	// Append multiple 0xFF bytes to ensure we include all documents with this value
+	return append(key, 0xFF, 0xFF, 0xFF, 0xFF)
```
[the remark this answers is not in the document]
Comment (fredcarle) [reply internal/db/fetcher/indexer_iterators.go:501]: question: Why is appending bytes better that incrementing the provided key?

File: internal/db/fetcher/indexer_iterators.go (hunk 3 above)
Comment (fredcarle) [reply internal/db/fetcher/indexer_iterators.go:379]: I just realized it's reflecting the underlying iterator configuration. It feels off but my brain is still not fully functional yet so I'll leave it at that for now 🙃😶‍🌫️.

File: internal/db/fetcher/indexer_iterators.go [hunk 5]
```diff
@@ -93,30 +103,57 @@ func (iter *indexPrefixIterator) Init(ctx context.Context, store datastore.DSRea
 		}
 	}
 	iter.resultIter = nil
+
+	var iterOpts corekv.IterOptions
+	if iter.prefixKey != nil {
```
[the remark this answers is not in the document]
Comment (islamaliev) [reply internal/db/fetcher/indexer_iterators.go:108]: done

File: internal/db/fetcher/indexer_iterators.go (hunk 3 above)
Comment (islamaliev) [reply internal/db/fetcher/indexer_iterators.go:379]: renamed

File: internal/db/fetcher/indexer_iterators.go (hunk 2 above)
Comment (islamaliev) [reply internal/db/fetcher/indexer_iterators.go:181]: renamed

File: internal/db/fetcher/indexer_iterators.go (hunk 1 above)
Comment (islamaliev) [reply internal/db/fetcher/indexer_iterators.go:261]: added a comment

File: internal/db/fetcher/indexer_iterators.go [hunk 6]
```diff
@@ -440,6 +493,147 @@ func (f *indexFetcher) newIndexDataStoreKeyWithValues(values []client.NormalValu
...
+			valueKey := f.createKeyWithValue(baseKey, cond.val)
+			endKey = incrementKeyBytes(valueKey.Bytes())
+		}
+	}
+
+	return startKey, endKey, nil
+}
+
+// isRangeCompatible checks if a filter condition is compatible with range queries.
+func (f *indexFetcher) isRangeCompatible(cond fieldFilterCond) bool {
```
[the remark this answers is not in the document]
Comment (islamaliev) [reply internal/db/fetcher/indexer_iterators.go:586]: I added them to ignore JSON for now. But now I added JSON as well

File: internal/db/fetcher/indexer_iterators.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (islamaliev) [reply internal/db/fetcher/indexer_iterators.go:501]: > todo: Please document if exactly 4 0xFFs is significant or arbitrary, and if it is significant please describe why.

You're right to question this - the comment is a bit misleading. The function creates an upper boundary for partial index keys used in range queries. For composite indexes, there could be additional field values after the range field, not just the document ID. The 4 0xFF bytes ensure the boundary is lexicographically greater than any possible completion of the key, whether that's more field values or the final document ID. The comment is updated

[the remark this answers is not in the document]
Comment (islamaliev) [reply internal/db/fetcher/indexer_iterators.go:501]: > Why is appending bytes better that incrementing the provided key?

Incrementing a byte array is complex and error-prone. You need to deal with overflow. What if the last byte is already 0xFF? Or [0xFF, 0xFF] incremented becomes [0x01, 0x00, 0x00] . Also when you change bytes directly you somewhat messing with bytes encoding which for some edge cases might lead to unpredictable behaviour.

File: internal/db/fetcher/indexer_iterators.go [hunk 7]
```diff
@@ -69,22 +69,32 @@ type indexIterResult struct {
 	value    []byte
 }
```
[the remark this answers is not in the document]
Comment (islamaliev) [reply internal/db/fetcher/indexer_iterators.go:71]: I added a ticket for this https://github.com/sourcenetwork/defradb/issues/3790

File: internal/db/fetcher/indexer_iterators.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (AndrewSisley) [reply internal/db/fetcher/indexer_iterators.go:510]: I feel like I understand this code less than I thought I did.

It feels like you are going to be overshooting your keys, and potentially reading other, unwanted keys from the store (or missing wanted values when approaching the end key) - anything between `key+1` and `key+(4*0xFF)`.

Consider:
```
if descending {
case opLt:
			// For descending index, we want values < X
			// Start just after X and go to the end
			valueKey := f.createKeyWithValue(baseKey, cond.val)
			startKey = incrementKeyBytes(valueKey.Bytes())
			endKey = incrementKeyBytes(baseKey.Bytes())
```
This would miss stuff lying between `valueKey+1` and `valueKey+(4*0xFF)`, and items between `baseKey+1` and `baseKey+(4*0xFF)`?

**Review by AndrewSisley — CHANGES_REQUESTED** [review]
I am confused and concerned about the `incrementKeyBytes` now

File: internal/db/fetcher/indexer_iterators.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (AndrewSisley) [reply internal/db/fetcher/indexer_iterators.go:510]: Ah of course - I see - you are relying on the assumption that there will be no index keys with a value of `/CollectionID/IndexID/25/0xFF0xFF0xFF0xFF`.  This might be incorrect no? It looks like it will work for where `incrementKeyBytes` is called now, as it is only called for gt(e)/lt(e) cases, which atm are only available for numeric types of fixed length.

It very much looks like it will break if we ever use this function for variable length types (string/blob/floatN/intN/etc).

suggestion: I have a fairly strong preference for it to actually increment the byte to the next prefix (`/CollectionID/IndexID/26/`), as whilst it is more complicated, it is always going to be correct, and is going to be much easier for readers to understand. 

If you really want to leave the compiled code as-is, please do the following:

todo: Paste your last reply into the function documentation with an large warning stating that this function will only work for the current cases.  It would be really easy to blindly use this function and forget (and not test) the edge cases I am concerned about.

todo: Please rename this function as it is not really incrementing bytes, but shifting/appending them.  The current name is misleading IMO.
